### PR TITLE
feat: Change trace struct

### PR DESCRIPTION
**Pull request by Lawliet-Chan**
Break change data structure `blockTrace`.
According to https://github.com/scroll-tech/go-ethereum/pull/181 and https://github.com/scroll-tech/go-ethereum/pull/184

### CODE REVIEW

**Comment by lispc** [comment]
can `make mock` and `make mock_pack` pass?

**Review by noel2004 — COMMENTED** [review]
Some original test case include mutiple txs in one block

For example, the `nft.json` has 3 txs for minting, transfer and burn respectively.

Though it seems to be ok if we test the new cases with mutiple block mode, I am not very sure if we would had missed something ...

**Comment by Lawliet-Chan** [comment]
```
running 1 test
[2022-11-23T10:10:13Z INFO  integration::test_util] git version prealpha-v2.0-25-gdbb9c47
[2022-11-23T10:10:13Z INFO  integration] test cases traces: ["tests/traces/bridge/01.json", "tests/traces/bridge/02.json", "tests/traces/bridge/03.json", "tests/traces/bridge/04.json", "tests/traces/bridge/05.json", "tests/traces/bridge/06.json", "tests/traces/bridge/07.json", "tests/traces/bridge/08.json", "tests/traces/bridge/09.json", "tests/traces/bridge/10.json", "tests/traces/bridge/11.json", "tests/traces/bridge/12.json", "tests/traces/bridge/13.json", "tests/traces/bridge/14.json", "tests/traces/bridge/15.json", "tests/traces/bridge/16.json", "tests/traces/dao_dao-Cancel.json", "tests/traces/dao_dao-Propose.json", "tests/traces/dao_deploy.json", "tests/traces/dao_gov-deploy.json", "tests/traces/erc20_deploy.json", "tests/traces/erc20_mint.json", "tests/traces/erc20_transfer.json", "tests/traces/greeter_deploy.json", "tests/traces/greeter_setValue.json", "tests/traces/nft_burn.json", "tests/traces/nft_deploy.json", "tests/traces/nft_mint.json", "tests/traces/nft_transferFrom.json", "tests/traces/sushi_approve.json", "tests/traces/sushi_chef-add.json", "tests/traces/sushi_chef-deploy.json", "tests/traces/sushi_chef-deposit.json", "tests/traces/sushi_chef-set.json", "tests/traces/sushi_chef-withdraw.json", "tests/traces/sushi_deploy.json", "tests/traces/sushi_mint.json", "tests/traces/sushi_transferOwnership.json", "tests/traces/uniswapv2_btc-deploy.json", "tests/traces/uniswapv2_factory-createPair.json", "tests/traces/uniswapv2_factory-deploy.json", "tests/traces/uniswapv2_initBalance.json", "tests/traces/uniswapv2_router-AddLiquidity.json", "tests/traces/uniswapv2_router02-deploy.json", "tests/traces/uniswapv2_weth9-deploy.json"]
[2022-11-23T10:10:13Z INFO  integration] test state circuit with tests/traces/bridge/01.json
[2022-11-23T10:10:13Z INFO  zkevm::prover] start mock prove state
[2022-11-23T10:10:13Z INFO  bus_mapping::circuit_input_builder] handling block Some(1)
[2022-11-23T10:10:15Z INFO  zkevm::prover] mock prove state done
[2022-11-23T10:10:15Z INFO  integration] test state circuit with tests/traces/bridge/01.json result: Ok(())
[2022-11-23T10:10:15Z INFO  integration] test state circuit with tests/traces/bridge/02.json
[2022-11-23T10:10:15Z INFO  zkevm::prover] start mock prove state
[2022-11-23T10:10:15Z INFO  bus_mapping::circuit_input_builder] handling block Some(2)
thread 'test_mock_prove_all_target_circuits' panicked at 'arithmetic operation overflow', /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/primitive-types-0.11.1/src/lib.rs:38:1
stack backtrace:
   0: rust_begin_unwind
             at /rustc/015a824f2dffe32707fceb59c47effaf7b73486c/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at /rustc/015a824f2dffe32707fceb59c47effaf7b73486c/library/core/src/panicking.rs:142:14
   2: <primitive_types::U256 as core::ops::arith::SubAssign>::sub_assign
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/uint-0.9.3/src/uint.rs:1429:5
   3: bus_mapping::circuit_input_builder::CircuitInputBuilder::handle_block_inner
             at /home/ubuntu/.cargo/git/checkouts/zkevm-circuits-462c2cba34f9b301/bd9b8ee/bus-mapping/src/circuit_input_builder.rs:183:17
   4: zkevm::circuit::builder::block_results_to_witness_block
             at ./src/circuit/builder.rs:82:9
   5: <zkevm::circuit::StateCircuit as zkevm::circuit::TargetCircuit>::from_block_results
             at ./src/circuit.rs:166:29
   6: zkevm::prover::Prover::mock_prove_target_circuit_multi
             at ./src/prover.rs:436:35
   7: zkevm::prover::Prover::mock_prove_target_circuit
             at ./src/prover.rs:428:9
   8: integration::test_mock_prove_all_with_circuit
             at ./tests/integration.rs:80:22
   9: integration::test_mock_prove_all_target_circuits
             at ./tests/integration.rs:133:30
  10: integration::test_mock_prove_all_target_circuits::{{closure}}
             at ./tests/integration.rs:115:1
  11: core::ops::function::FnOnce::call_once
             at /rustc/015a824f2dffe32707fceb59c47effaf7b73486c/library/core/src/ops/function.rs:248:5
  12: core::ops::function::FnOnce::call_once
             at /rustc/015a824f2dffe32707fceb59c47effaf7b73486c/library/core/src/ops/function.rs:248:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
test test_mock_prove_all_target_circuits ... FAILED

```

I met a problem when I `make mock` as above. (Any other proving also causes this bug.)
Could you help me take a look? I guess it is related to the code I changed but now I cannot figure out what's wrong.  How can I fix it.  @lispc @noel2004 @HAOYUatHZ  


BTW, from the log above,  the process panics here: 

> at /home/ubuntu/.cargo/git/checkouts/zkevm-circuits-462c2cba34f9b301/bd9b8ee/bus-mapping/src/circuit_input_builder.rs:183:17

**Comment by lispc** [comment]
cannot guess the reason quickly. So my suggestion is to: 

(1) on main branch, print the struct passed to `block_results_to_witness_block` to a file
(2) on this branch, print too.
(3) then open two log file to check inputs. It is very possible that the new struct misses some useful fields.

**Comment by Lawliet-Chan** [comment]
> cannot guess the reason quickly. So my suggestion is to:
> 
> (1) on main branch, print the struct passed to `block_results_to_witness_block` to a file (2) on this branch, print too. (3) then open two log file to compare inputs. It is very possible that the new struct misses some useful fields.

What is passed to `block_results_to_witness_block` is just `block_trace` and it is read from `zkevm/tests/traces/`. And we just break change the block-trace structure. https://github.com/scroll-tech/go-ethereum/pull/181

I think we can just compare the trace files of this branch with the main branch.

**Comment by lispc** [comment]
i find 02.json is full of native eth transfer. so another suggestion:

skip 02.json, test 03, 04... If normal contract call runs ok and only native transfer fails, then check native transfer handling more carefully.

File: types/src/eth.rs [hunk 1]
```diff
@@ -5,18 +5,19 @@ use mpt_circuits::serde::SMTTrace;
 use serde::{Deserialize, Serialize};
 use std::collections::HashMap;
 
-/// BlockResultWrapper is the payload from Scroll.
+/// TaskMsg is the payload from Scroll.
 #[derive(Deserialize, Serialize, Default, Debug)]
-pub struct BlockResultWrapper {
+pub struct TaskMsg {
     pub id: u64,
```
Comment (HAOYUatHZ) [inline types/src/eth.rs:11]: r u sure this is `u64`?

File: types/src/eth.rs [hunk 2]
```diff
@@ -25,6 +26,36 @@ pub struct BlockResult {
...
+}
+
+#[derive(Deserialize, Serialize, Default, Debug, Clone)]
+pub struct TransactionData {
+    #[serde(rename = "isCreate", default)]
+    pub is_create: bool,
+    pub from: Option<Address>,
+    #[serde(flatten)]
+    pub transaction: Transaction,
+}
```
Comment (HAOYUatHZ) [inline types/src/eth.rs:57]: we can replace this whole struct with `Transaction` now

Comment (Lawliet-Chan) [reply types/src/eth.rs:57]: Do we need to change the l2geth first?

**Comment by HAOYUatHZ** [comment]
> It is very possible that the new struct misses some useful fields.

The new struct only add more fields. I explained to @Lawliet-Chan that he probably forgot to manually assign some needed data.

**Comment by Lawliet-Chan** [comment]
When I use new traces from @HAOYUatHZ , it still failed.  `make mock` will causes: 
> running 1 test
[2022-11-24T13:15:55Z INFO  integration::test_util] git version prealpha-v2.0-40-g087314a
[2022-11-24T13:15:55Z INFO  integration] test cases traces: ["tests/traces/bridge/01.json", "tests/traces/bridge/02.json", "tests/traces/bridge/03.json", "tests/traces/bridge/04.json", "tests/traces/bridge/05.json", "tests/traces/bridge/06.json", "tests/traces/bridge/07.json", "tests/traces/bridge/08.json", "tests/traces/bridge/09.json", "tests/traces/bridge/10.json"]
[2022-11-24T13:15:55Z INFO  integration] test state circuit with tests/traces/bridge/01.json
[2022-11-24T13:15:55Z INFO  zkevm::prover] start mock prove state
[2022-11-24T13:15:55Z INFO  bus_mapping::circuit_input_builder] handling block Some(1)
[2022-11-24T13:15:59Z INFO  zkevm::prover] mock prove state done
[2022-11-24T13:15:59Z INFO  integration] test state circuit with tests/traces/bridge/01.json result: Ok(())
[2022-11-24T13:15:59Z INFO  integration] test state circuit with tests/traces/bridge/02.json
[2022-11-24T13:15:59Z INFO  zkevm::prover] start mock prove state
[2022-11-24T13:15:59Z INFO  bus_mapping::circuit_input_builder] handling block Some(2)
[2022-11-24T13:16:03Z INFO  zkevm::prover] mock prove state done
[2022-11-24T13:16:03Z INFO  integration] test state circuit with tests/traces/bridge/02.json result: Ok(())
[2022-11-24T13:16:03Z INFO  integration] test state circuit with tests/traces/bridge/03.json
[2022-11-24T13:16:03Z INFO  zkevm::prover] start mock prove state
[2022-11-24T13:16:03Z INFO  bus_mapping::circuit_input_builder] handling block Some(3)
[2022-11-24T13:16:06Z INFO  zkevm::prover] mock prove state done
[2022-11-24T13:16:06Z INFO  integration] test state circuit with tests/traces/bridge/03.json result: Ok(())
[2022-11-24T13:16:06Z INFO  integration] test state circuit with tests/traces/bridge/04.json
[2022-11-24T13:16:06Z INFO  zkevm::prover] start mock prove state
[2022-11-24T13:16:06Z INFO  bus_mapping::circuit_input_builder] handling block Some(4)
[2022-11-24T13:16:08Z INFO  zkevm::prover] mock prove state done
[2022-11-24T13:16:08Z INFO  integration] test state circuit with tests/traces/bridge/04.json result: Ok(())
[2022-11-24T13:16:08Z INFO  integration] test state circuit with tests/traces/bridge/05.json
[2022-11-24T13:16:08Z INFO  zkevm::prover] start mock prove state
[2022-11-24T13:16:08Z INFO  bus_mapping::circuit_input_builder] handling block Some(5)
[2022-11-24T13:16:08Z INFO  bus_mapping::circuit_input_builder] handling 0th(inner idx: 0) tx 0x3c0d31720a81abe31a97d49cd76a929d63d7a442a3725a488f4008867491ffc1
[2022-11-24T13:16:08Z INFO  integration] test state circuit with tests/traces/bridge/05.json result: Err(AccountNotFound(0x16e77e688217d83003a830fdb7886865b70e90de))
[2022-11-24T13:16:08Z INFO  integration] test state circuit with tests/traces/bridge/06.json
[2022-11-24T13:16:08Z INFO  zkevm::prover] start mock prove state
[2022-11-24T13:16:08Z INFO  bus_mapping::circuit_input_builder] handling block Some(6)
[2022-11-24T13:16:08Z INFO  bus_mapping::circuit_input_builder] handling 0th(inner idx: 0) tx 0xfd720c260921dad035e050d7e13dafb11473c2bc89e34da655bf296fa5a04a1e
[2022-11-24T13:16:08Z INFO  integration] test state circuit with tests/traces/bridge/06.json result: Err(AccountNotFound(0x5c03093cd583f984dc9788ce9b8aa9a730c9955f))
[2022-11-24T13:16:08Z INFO  integration] test state circuit with tests/traces/bridge/07.json
[2022-11-24T13:16:08Z INFO  zkevm::prover] start mock prove state
[2022-11-24T13:16:08Z INFO  bus_mapping::circuit_input_builder] handling block Some(7)
[2022-11-24T13:16:08Z INFO  bus_mapping::circuit_input_builder] handling 0th(inner idx: 0) tx 0xbd1b89b2bda1b138d73e49438b55f87d03db8500c3755888cbbc1dfd336a5979
[2022-11-24T13:16:08Z INFO  integration] test state circuit with tests/traces/bridge/07.json result: Err(AccountNotFound(0x7ca349863061d2edeeacc1b0c917b1d9b510f9d4))
[2022-11-24T13:16:08Z INFO  integration] test state circuit with tests/traces/bridge/08.json
[2022-11-24T13:16:08Z INFO  zkevm::prover] start mock prove state
[2022-11-24T13:16:08Z INFO  bus_mapping::circuit_input_builder] handling block Some(8)
[2022-11-24T13:16:08Z INFO  bus_mapping::circuit_input_builder] handling 0th(inner idx: 0) tx 0xbc606cdada7a262ece9bd3be8c888e272e9017c430ff0b6b4a739a7506284b03
thread 'test_mock_prove_all_target_circuits' panicked at 'arithmetic operation overflow', /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/primitive-types-0.11.1/src/lib.rs:38:1
stack backtrace:
   0: rust_begin_unwind
             at /rustc/015a824f2dffe32707fceb59c47effaf7b73486c/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at /rustc/015a824f2dffe32707fceb59c47effaf7b73486c/library/core/src/panicking.rs:142:14
   2: bus_mapping::evm::opcodes::gen_end_tx_ops
   3: bus_mapping::circuit_input_builder::CircuitInputBuilder::handle_block_inner
   4: zkevm::circuit::builder::block_results_to_witness_block
   5: <zkevm::circuit::StateCircuit as zkevm::circuit::TargetCircuit>::from_block_results
   6: zkevm::prover::Prover::mock_prove_target_circuit_multi
   7: core::ops::function::FnOnce::call_once
   8: core::ops::function::FnOnce::call_once
             at /rustc/015a824f2dffe32707fceb59c47effaf7b73486c/library/core/src/ops/function.rs:248:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
test test_mock_prove_all_target_circuits ... FAILED



I use `prove tool` proved failed too except `01.json~04.json`, too.   But 01.json~04.json are all empty blocks (They have no transactions)

**Comment by HAOYUatHZ** [comment]
format those 01~10.json files

File: types/src/eth.rs [hunk 3]
```diff
@@ -5,18 +5,19 @@ use mpt_circuits::serde::SMTTrace;
...
-/// BlockResultWrapper is the payload from Scroll.
+/// TaskMsg is the payload from Scroll.
 #[derive(Deserialize, Serialize, Default, Debug)]
-pub struct BlockResultWrapper {
-    pub id: u64,
+pub struct TaskMsg {
+    pub id: String,
     #[serde(rename = "blockTraces")]
-    pub block_result: BlockResult,
+    pub block_result: Vec<BlockTrace>,
```
[the remark this answers is not in the document]
Comment (Lawliet-Chan) [reply types/src/eth.rs]: I found that `TaskMsg` is the struct between `coordinator` and `roller`.   `common-rs` do not care about it, so I remove it.

File: zkevm/src/circuit.rs [hunk 4]
```diff
@@ -32,11 +32,11 @@ pub trait TargetCircuit {
     /// used to generate vk&pk
     fn empty() -> Self::Inner;
     //fn public_input_len() -> usize { 0 }
-    fn from_block_result(block_result: &BlockResult) -> anyhow::Result<(Self::Inner, Vec<Vec<Fr>>)>
+    fn from_block_result(block_result: &BlockTrace) -> anyhow::Result<(Self::Inner, Vec<Vec<Fr>>)>
```
Comment (HAOYUatHZ) [inline zkevm/src/circuit.rs:35]: rename block_result(s) to block_trace(s) in 
+ zkevm/src/circuit.rs
+ zkevm/src/circuit/builder.rs
+ zkevm/src/prover.rs
+ zkevm/src/utils.rs
+ zkevm/tests/integration.rs

**Comment by lispc** [comment]
ping me when the branch is ready @Lawliet-Chan

**Comment by Lawliet-Chan** [comment]
> ping me when the branch is ready @Lawliet-Chan

Ready now. @lispc @HAOYUatHZ 
I've already tested pass by `make mock` and `make mock_pack` with the traces of `/bridge` , `/dao` , `/erc20`, `/nft`, `/sushi`

**Review by HAOYUatHZ — APPROVED** [review]

**Review by lispc — APPROVED** [review]

**Comment by HAOYUatHZ** [comment]
Looks good. I will update these in the future.
image

File: zkevm/tests/test_util.rs [hunk 5]
```diff
@@ -20,14 +20,13 @@ pub fn parse_trace_path_from_mode(mode: &str) -> &'static str {
...
-        "single" => "./tests/traces/single-erc20.json",
-        "single_legacy" => "./tests/traces/single-erc20-legacy.json",
-        "dao" => "./tests/traces/dao.json",
-        "nft" => "./tests/traces/nft.json",
-        "sushi" => "./tests/traces/masterchef.json",
-        _ => "./tests/traces/multiple-erc20.json",
+        "single" => "./tests/traces/erc20/single.json",
+        "multiple" => "./tests/traces/erc20/multiple.json",
+        "native" => "./tests/traces/native_transfer.json",
+        "dao" => "./tests/traces/dao/propose.json",
```
Comment (noel2004) [inline zkevm/tests/test_util.rs:26]: So here we change the test sample from including mutiple actions (propose and cancel) to single one. I am not sure if this degration is appropriate

**Review by noel2004 — COMMENTED** [review]
Maybe we should use mutiple block mode and test mutiple trace files in each mode. I can try updating this part sometime later

**Review by noel2004 — APPROVED** [review]

File: zkevm/tests/test_util.rs (hunk 5 above)
Comment (HAOYUatHZ) [reply zkevm/tests/test_util.rs:26]: I think it's fine.

**Pull request merged**